tracks/julia/exercises/nucleotide-count/mentoring.md

```diff
@@ -0,0 +1,40 @@
+### Mentor guidance
+
+Students that use four counting variables rather than mutating a dict should be shown the dict solution or asked to come up with it on their own if there is some doubt that they knew they could do that.
+
+We shouldn't require the mutating-a-dict strategy because it's actually pretty slow.
+
+
+### Example solutions
+
+```julia
+function count_nucleotides(strand)
+    counter = Dict(base => 0 for base in "ACGT")
```

Comment:
That's a really nice way of constructing this dict, haven't seen that before :D

Reply:
I would rather *not* show the Dict solution - most of my time mentoring these past two weeks has been explaining that `Dict` or `Set` is often *not* their friend, if they care at all about performance.

Reply:
Do you have an example of a solution without Dicts that could be added here?

Reply:
The reason for this is that I would like the students to know how to use dictionaries and index collections.

I generally encourage students to give correct, concise answers and not to worry too much about performance, and I think that's a decent default for real-world programming, or at least it has been for me.

Scott, you have more experience than me, though :)

I think we should show the more imperative/explicit fast solutions, too, though. Something for these after-exercise markdown files that we're getting in v3 (for all exercises or just concept?)

Reply:
I think that exercises where a `Dict` is the optimal solution, on all fronts, should be added then.
Many people try out Julia precisely because of they've heard about the performance, and if they use non-performant techniques, and then see that it's not any faster than Python, then they will no longer be interested in pursuing it.
(That almost happened to me, when I saw how much slower Julia was for string handling!)

In my background, performance has always been critical - it meant the difference between people choosing our solution, or the competitions.
I've seen projects (and companies) die because not enough care was taken about performance.

I'd encourage students to first come up with a clean, easy to read solution, but never to stop just there, show them where that might end up with O(N) or O(N^2) algorithms, and how to improve on that.

Reply:
> Do you have an example of a solution without Dicts that could be added here?

Yes, I'll do so tomorrow or Monday!

Reply:
I wrote one last night and added it :)

Reply:
Well, we still construct a dict right at the end, but that could be a named tuple if the spec was changed.

Reply:
Yes, I think using a named tuple would be very nice (showing off an important feature of Julia also)